torch/csrc/dynamo/guards.cpp

```diff
@@ -0,0 +1,378 @@
+    return NULL;
+  }
+
+  std::vector<std::string> tensor_check_names;
+  tensor_check_names.reserve(names_size);
+  for (auto i : c10::irange(names_size)) {
+    PyObject* value = PyList_GetItem(tensor_check_names_py, i);
+    if (!PyUnicode_Check(value)) {
+      PyErr_SetString(
+          PyExc_TypeError, "tensor_check_names must only contain strings");
+      return NULL;
+    }
+    tensor_check_names.emplace_back(PyUnicode_AsUTF8(value));
+  }
+
+  LocalState state;
+  for (auto i : c10::irange(len)) {
+    PyObject* item = PyTuple_GET_ITEM(args, i);
+    if (Py_TYPE(item) != checks[i].pytype) {
+      std::stringstream fail_reason;
+      PyObject* type_str = PyObject_Str(PyObject_Type(item));
+      fail_reason << "expected type of '" << tensor_check_names[i]
+                  << "' to be a tensor type, ";
+      if (!type_str) {
+        fail_reason << "but found a different type";
+      } else {
+        fail_reason << "' but found " << PyUnicode_AsUTF8(type_str);
+      }
+      return Py_BuildValue("s", fail_reason.str().c_str());
+    }
+    std::string fail_reason = checks[i].check_verbose(
+        state, THPVariable_Unpack(item), tensor_check_names[i]);
+    if (fail_reason.length() > 0) {
+      return Py_BuildValue("s", fail_reason.c_str());
+    }
+  }
+
+  Py_RETURN_TRUE;
+}
+
+static PyMethodDef TensorGuards_methods[] = {
+    {"check", (PyCFunction)TensorGuards_check, METH_VARARGS, ""},
+    {"check_verbose",
+     (PyCFunction)(void*)TensorGuards_check_verbose,
+     METH_VARARGS | METH_KEYWORDS,
+     "verbose fail reasons for failed checks"},
+    {NULL} /* Sentinel */
+};
+
+static PyTypeObject TensorGuardsType = {
+    // NOLINTNEXTLINE
+    PyVarObject_HEAD_INIT(NULL, 0)};
+
+static PyObject* check_type_id(PyObject* dummy, PyObject* args) {
+  // faster `lambda obj, expected: id(type(obj)) == expected`
+  PyObject* obj;
+  unsigned long expected;
+  if (!PyArg_ParseTuple(args, "Ok", &obj, &expected)) {
+    return NULL;
+  }
+  if (Py_TYPE(obj) == (void*)expected) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+}
+
+static PyObject* check_obj_id(PyObject* dummy, PyObject* args) {
+  // faster `lambda obj, expected: id(obj) == expected`
+  PyObject* obj;
+  unsigned long expected;
+  if (!PyArg_ParseTuple(args, "Ok", &obj, &expected)) {
+    return NULL;
+  }
+  if (obj == (void*)expected) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+}
+
+static PyMethodDef _methods[] = {
+    {"check_type_id", check_type_id, METH_VARARGS, NULL},
+    {"check_obj_id", check_obj_id, METH_VARARGS, NULL},
+    {NULL, NULL, 0, NULL}};
+
+static struct PyModuleDef _module = {
+    PyModuleDef_HEAD_INIT,
+    "torch._C._dynamo.guards",
+    "Module containing checks on tensors",
+    -1,
+    _methods};
+
+} // namespace
+
+PyObject* torch_c_dynamo_guards_init() {
+  // initialize TensorGuardsType
+  TensorGuardsType.tp_name = "torchdynamo._guards.TensorGuards";
```

Comment:
Are we going to keep the `torchdynamo` top level module? Or should this be in `torch._dynamo`.

Reply:
Good catch, will fix. This is actually `torch._C._dynamo.guards.TensorGuards` now.